### ISSUE TITLE
Postconditions on user-defined conversion parameters are not respected

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#49575]
**Version Used**: dc6f316a5e4f8eb3a0d36ab356c5c12686db00a9

**Steps to Reproduce**:

```cs
#nullable enable

using System.Diagnostics.CodeAnalysis;

public static class Program
{
    public static void M1(string? s)
    {
        AssertNotNull(s);
        s.ToString();
    }

    public static void M2(string? s)
    {
        C c = s;
        s.ToString(); // 1
    }

    public static void AssertNotNull([AllowNull, NotNull] string s) { throw null!; }

    public class C
    {
        public static implicit operator C([AllowNull, NotNull] string s) { throw null!; }
    }
}
```

[SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4GIB2BXANliwtwAEcaeBAsAFBUACADITQIwB0AIgJYQDmaA9rA4BjKCwDCfACZwAgqSwBPKBygBuKrQBMAZioBvKoSONtjJgDZGAFkIBZJgApmdAPyEoASkPGDlY/8IZKChEGAA5PnDsLAdPdT8Ao1EAFT4AZRgEDjRuBw94/wBfDQSjGlNmSxobW00nJld3L1LCX0SjMUIhQgBedwL2lPTM7Nz8wgB6CcImbyNi6hbysyqbIJCEcMiw6IcAbRkcPgB3HZwAGkIIqJwAXTMGT1bCGAALBBPCTBwAQlVCBZzEyMTSEMRAtrtZaVQgcMAABywwg4MEIfHhiAgMD4CDB+0OWBOZywl2uxPuziazzeH2OX2ifwBQIWCyAA=)

**Expected Behavior**: No diagnostics

**Actual Behavior**: Diagnostic indicated by `// 1`.
